3.0.0.md

```diff
@@ -0,0 +1,209 @@
+# Upgrading Parse Server to version 3.0.0
+
+With the 3.0.0, parse-server has completely revamped it's cloud code interface. Gone are the backbone style calls, welcome promises and async await.
+
+In order to leverage those new nodejs features, you'll need to run at least node 8.10. We've dropped the support of node 6 a few months ago, so you should feed really comfortable at home.
+
+## Migrating Cloud Code Hooks
+
+### Synchronous validations:
+
+```js
+// before
+Parse.Cloud.beforeSave('MyClassName', function(request, response) {
+  if (!passesValidation(req.object)) {
+    response.error('Ooops something went wrong');
+  } else {
+    response.success();
+  }
+});
+
+// after
+Parse.Cloud.beforeSave('MyClassName', (request) => {
+  if (!passesValidation(req.object)) {
+    throw 'Ooops something went wrong'; 
+  }
+});
+```
+
+All methods are wrapped into promises, so you can freely `throw` `Error`, `Parse.Error` or `string` to mark an error.
+
+### Asynchronous validations:
+
+For asynchronous code, you can use promises or async / await.
+
+Consider the following beforeSave call that would replace the contents of a fileURL with a proper copy of the image as a Parse.File.
+
+```js
+// before
+Parse.Cloud.beforeSave('Post', function(request, response) {
+  Parse.Cloud.httpRequest({
+    url: request.object.get('fileURL'),
+    success: function(contents) {
+      const file = new Parse.File('image.png', { base64: contents.buffer.toString('base64') });
+      file.save({
+        success: function() {
+          request.object.set('file', file);
+          response.success();
+        },
+        error: response.error
+      });
+    }, 
+    error: response.error
+  });
+});
+```
+
+As we can see the current way, with backbone style callbacks is quite tough to read and maintain.
+It's also not really trivial to handle errors, as you need to pass the response.error to each error handlers.
+
+Now it can be modernized with promises:
+
+```js
+// after (with promises)
+Parse.Cloud.beforeSave('MyClassName', (request) => {
+  return Parse.Cloud.httpRequest({
+    url: request.object.get('fileURL')
+  }).then((response) => {
+    const file = new Parse.File('image.png', { base64: contents.buffer.toString('base64') });
+    request.object.set('file', file);
+    return file.save();
+  });
+});
+```
+
+And you can even make it better with async await.
+
+```js
+// after with async/await
+Parse.Cloud.beforeSave('MyClassName', async (request) => {
+  const contents = await Parse.Cloud.httpRequest({
+    url: request.object.get('fileURL')
+  });
+
+  const file = new Parse.File('image.png', { base64: contents.buffer.toString('base64') });
+  request.object.set('file', file);
+  await file.save();
+});
+```
+
+Note: whenever a failure will occur, the beforeSave will fail.
```

Comment:
Note: if an error occurs or a string is thrown, the beforeSave will fail to save the state of the object being saved.